### DISCUSSION TITLE
Feature Request: allow control characters as column and row delimiters in read_csv

### DISCUSSION THREAD

**Discussion by kspeeckaert** [duckdb/duckdb#15049] [Ideas]
This is based on the blog post [CSVs Are Kinda Bad. DSVs Are Kinda Good.](https://matthodges.com/posts/2024-08-12-csv-bad-dsv-good/) (I'm not the author).

In an attempt to avoid the usual shenanigans with CSV delimiters, escaping, etc, we've decided to use the ASCII control characters _record separator_ (ASCII 30) and _unit separator_ (ASCII 31) as resp. line and column delimiters. This makes processing less error prone, but so far I haven't been able to convince `read_csv` to properly parse these (or maybe I've been doing it wrong). 

I do have a workaround available, using `read_text` and then some list-foo magic, which works, but I'm secretly hoping it's trivial to add to the existing code base, which would make DuckDB even more awesome than it already is!

As for my workaround, here it is:

Macro definition:
```sql
CREATE OR REPLACE MACRO read_dsv(path_spec)
 AS TABLE
(
with _lines as (
    select 
        filename
        ,regexp_split_to_array(content, '\x1e') as content
    from read_text(path_spec )
)
, _cleanup as (
    select 
        filename
        ,regexp_split_to_array(content[1],'\x1f') as header
        ,generate_subscripts(content[2:],1) as linenum
        ,unnest((content[2:]).list_filter(x -> trim(x) != '').list_transform(x -> x.regexp_split_to_array('\x1f'))) as line
    from _lines
)
select
    filename
    ,linenum
    ,unnest(map_entries(map(header, line)), recursive := true) as kv
from _cleanup
);
```
Usage:
```sql
-- You can use the same path specification as you would with read_text or read_csv, this includes globbing.
-- Trying to include the pivot statement in the macro isn't possible, as you then have to explicitly define the column values (which defeats the purpose of this implementation)
pivot read_dsv("C:\Temp\csv\*.csv")
on key
using first(value)
group by filename, linenum
order by filename, linenum
```

**Comment by Tishj** [comment]
Perhaps the newline parameter can be changed to accept this asan option.
I think that's all that has to be changed to support this right?

I'm not sure how much effort making that change would be though, iirc the newline options are only LF and CRLF currently

**Reply by kspeeckaert** [reply]
That's correct, basically this should work:

```sql
select * 
from read_csv("C:\Temp\csv\*.csv", delim=e'\x1f', new_line=e'\x1e') 
```
but the allowed options for `new_line` are '\r','\n', or '\r\n'. (corresponding [code reference](https://github.com/duckdb/duckdb/blob/9e6fd53f3dcf41b554740ed8eb62c8e41e0fdc43/src/execution/operator/csv_scanner/util/csv_reader_options.cpp#L174))

**Comment by soerenwolfers** [comment]
Just use parquet. Every sixth  DuckDB issues is a CSV issue. Why anybody would consciously decide to use CSV for anything, in particular anything new, is beyond me; choosing to make up your own DSV with a guarantee to run into even more problems with every single piece of tooling out there is even more bananas.
